File: lessons/compiled_lang/cpp/src/e50.branch.cpp
#include <cctype>
#include <iostream>

int main() {
    std::cout << "Input a character: ";
    char keypress;
    std::cin.get(keypress);

    // switch requires a single comparable value, not a pattern, so
    //  classify first and switch on the result
    int kind;
    if (std::isupper(static_cast<unsigned char>(keypress)))
        kind = 0;
    else if (std::islower(static_cast<unsigned char>(keypress)))
        kind = 1;
    else if (std::isdigit(static_cast<unsigned char>(keypress)))
        kind = 2;
    else
        kind = 3;

    switch (kind) {
        case 0: std::cout << "Uppercase letter" << std::endl; break;
        case 1: std::cout << "Lowercase letter" << std::endl; break;
        case 2: std::cout << "Digit" << std::endl; break;
        default: std::cout << "Punctuation, whitespace, or other" << std::endl;
    }

    return 0;
}

File: lessons/compiled_lang/cs/src/e50.branch.cs
using System;

class E50Branch
{
    static void Main()
    {
        Console.Write("Input a character: ");
        char keypress = (char)Console.Read();

        // pattern-matching switch (C# 7+) - `when` guards let each case
        //  test a predicate instead of just an equality constant
        switch (keypress)
        {
            case char c when char.IsUpper(c):
                Console.WriteLine("Uppercase letter");
                break;
            case char c when char.IsLower(c):
                Console.WriteLine("Lowercase letter");
                break;
            case char c when char.IsDigit(c):
                Console.WriteLine("Digit");
                break;
            default:
                Console.WriteLine("Punctuation, whitespace, or other");
                break;
        }
    }
}
